Add Neo4j extension

michael-simons opened the pull request:
This adds a new extension for the Neo4j Java Driver (see https://github.com/neo4j/neo4j-java-driver) based on the lastest 2.0 alpha version, to be released later this year.

The extension copies a lot from the netty extension, as the Neo4j driver shades Netty. Apart from that, the deployment module has to indicate that we need SSL in native mode.

The runtime module provides a basic configuration support in form of Neo4jDriverConfiguration, that allows configuring authentication of the driver. The configuration has some head room to configure additional aspects, as the similar aspect of the Spring Boot extension does (see https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/blob/master/neo4j-java-driver-spring-boot-autoconfigure/src/main/java/org/neo4j/driver/springframework/boot/autoconfigure/Neo4jDriverProperties.java).

The actual driver instance will be provided eagerly by a producer.

Usage is demonstrated in the integration tests, see io.quarkus.it.neo4j.Neo4jResource. I have enabled tests for blocking and asynchronous access, for testing reactive access we will have to wait for the public release of the Neo4j database 4.0 milestone later this month. With that in place, this extension will provide native, reactive database access for Neo4j on Quarkus.

The integration tests will run on the JVM with `mvn clean install -Pdocker-neo4j -Ptest-neo4j` and natively with `mvn clean install -Pdocker-neo4j -Ptest-neo4j`, in both cases an installed and working Docker environment are required.

michael-simons committed:
Add Neo4j extension.

This adds a new extension for the Neo4j Java Driver (see https://github.com/neo4j/neo4j-java-driver) based on the lastest 2.0 alpha version, to be released later this year.

The extension copies a lot from the netty extension, as the Neo4j driver shades Netty. Apart from that, the deployment module has to indicate that we need SSL in native mode. It will also run with `quarkus.ssl.native=false` natively, but must issue a warning in such case and disable encrypted client-server communication.

The runtime module provides a basic configuration support in form of Neo4jDriverConfiguration, that allows configuring authentication of the driver. The configuration has some head room to configure additional aspects, as the similar aspect of the Spring Boot extension does (see https://github.com/neo4j/neo4j-java-driver-spring-boot-starter/blob/master/neo4j-java-driver-spring-boot-autoconfigure/src/main/java/org/neo4j/driver/springframework/boot/autoconfigure/Neo4jDriverProperties.java).

The actual driver instance will be provided eagerly by a producer.

Usage is demonstrated in the integration tests, see io.quarkus.it.neo4j.Neo4jResource. I have enabled tests for blocking and asynchronous access, for testing reactive access we will have to wait for the public release of the Neo4j database 4.0 milestone later this month. With that in place, this extension will provide native, reactive database access for Neo4j on Quarkus.

The integration tests will run on the JVM with `mvn clean install -Dtest-neo4j -Ddocker` and natively with `mvn clean install -Dtest-neo4j -Ddocker -Dnative`, in both cases an installed and working Docker environment are required.

Co-Authored-By: Guillaume Smet <guillaume.smet@gmail.com>

michael-simons commented:
If it’s decided to accept and merge this, we should wait at least until Neo4j driver 2.0 alpha03 is released, as alpha02 has shaded Project reactor under its original package which is of course wrong and has been fixed in alpha03.

gsmet commented on integration-tests/neo4j/pom.xml:
```diff
@@ -0,0 +1,248 @@
+                                    <cleanupServer>true</cleanupServer>
+                                    <enableHttpUrlHandler>true</enableHttpUrlHandler>
+                                    <graalvmHome>${graalvmHome}</graalvmHome>
+                                    <enableJni>true</enableJni>
```
Do you really need JNI?

michael-simons replied:
Nope. That was a leftover from trying to make things work. Good catch.

michael-simons replied:
Update: Not working without. But, you have it in your docs, too: https://quarkus.io/guides/native-and-ssl-guide

```
-H:EnableURLProtocols=http,https --enable-all-security-services -H:+JNI
```

gsmet replied:
Did you try to remove the line or did you set it to false?

Because if you don't set it, the fact that your extension requires SSL should set it for you. Or do you need JNI even when SSL is disabled?

gsmet commented on integration-tests/neo4j/pom.xml:
```diff
@@ -0,0 +1,248 @@
+                                    <graalvmHome>${graalvmHome}</graalvmHome>
+                                    <enableJni>true</enableJni>
+                                    <debugBuildProcess>false</debugBuildProcess>
+                                    <additionalBuildArgs>--enable-https,-H:+ReportExceptionStackTraces</additionalBuildArgs>
```
We can't do SSL for native image on CI for now.

`-H:+ReportExceptionStackTraces` should be the default now so I think you can remove it.

gsmet commented on integration-tests/neo4j/src/main/resources/application.properties:
```diff
@@ -0,0 +1,3 @@
+quarkus.neo4j.uri = ${neo4j.uri}
+quarkus.neo4j.authentication.username = ${neo4j.username}
+quarkus.neo4j.authentication.password = ${neo4j.password}
```
We would need `quarkus.ssl.native=false` to get it to run properly on CI with native image.

Hopefully, it's gonna work without SSL?

michael-simons replied:
I now have a check in place. When native ssl is disabled (`!SslContextConfiguration.isSslNativeEnabled()`), than I turn off bolt encryption altogether. I added a note to the readme, though, because we really don't recommend that.

Worked in my tests, please check on CI.

gsmet commented on extensions/neo4j/deployment/src/main/java/io/quarkus/neo4j/deployment/NettyProcessor.java:
```diff
@@ -0,0 +1,101 @@
+        runtimeBeanBuildItemBuildProducer.produce(RuntimeBeanBuildItem.builder(EventLoopGroup.class)
+                .setSupplier(worker)
+                .setScope(ApplicationScoped.class)
+                .build());
```
Are these used?

michael-simons replied:
Yes. They have been copied from the Netty extension and do apply to the driver as well.

gsmet commented on extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/BossGroup.java:
```diff
@@ -0,0 +1,14 @@
+ */
+@Qualifier
+@Retention(RetentionPolicy.RUNTIME)
+public @interface BossGroup {
```
I'm not sure we need this?

michael-simons replied:
See above, as our driver shades Netty, all it's extensions and mods are applicable as well.

michael-simons replied:
I do have a dependency, see

`io.quarkus.neo4j.deployment.NettyProcessor`

It's an adapted version of the `NettyProcessor`, that has been adapted to use the shaded packages.

gsmet replied:
Sure. But do you use the beans generated by `NettyProcessor` somewhere? I don't see them consumed in your work but, as I said, I don't know Netty well so I might miss how these things are wired.

michael-simons replied:
I just removed it and it still works… For whatever reason, I ended up in a situation before where things didn't work in the end. Must have been confused by too many poms and files open exception. They will be gone in a sec. Thanks for not letting it go :)

gsmet reviewed (changes requested):
@michael-simons nice work.

I added some comments/questions inline. Could you take care of them?

You said you wanted us to wait for the alpha03 release. When is it planned?

gsmet commented:
@michael-simons I saw you adjusted a few things but not everything.

Can you ping me when you need another round of review? Thanks!

michael-simons commented:
Yes, would have done so and will do. It was a bit late, otherwise I wouldn’t have pushed things premature. Thanks for your input so far!

michael-simons commented on extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/runtime/Neo4jDriverConfiguration.java:
```diff
@@ -0,0 +1,45 @@
+ * @author Michael J. Simons
+ */
+@ConfigRoot(phase = ConfigPhase.RUN_TIME, name = "neo4j")
+public class Neo4jDriverConfiguration {
```
Good point.

michael-simons commented on extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/runtime/Neo4jDriverTemplate.java:
```diff
@@ -0,0 +1,51 @@
+    public void configureNeo4jProducer(BeanContainer beanContainer, Neo4jDriverConfiguration configuration,
+            ShutdownContext shutdownContext) {
+
+        if (driver == null) {
```
Better safe than sorry, I guess. I'm always trying to be as defensive as possible in case of relatively new territories.

gsmet replied:
In theory, I would tend to agree but here it makes things more complicated for no reason.

The template method will be called only once if you call it only once in your processor.

michael-simons commented on extensions/neo4j/runtime/src/main/java/io/quarkus/neo4j/runtime/Neo4jDriverTemplate.java:
```diff
@@ -0,0 +1,51 @@
+            authToken = AuthTokens.none();
+        } else {
+            uri = configuration.uri;
+            authToken = Optional.ofNullable(configuration.authentication)
```
The `username` is defaulted, so I think checking for it is also necessary.

michael-simons replied:
I have settled now with an additional `disabled` field on `authentication`, as `@ConfigGroups` cannot be made optional:

>  <em>Configuration group properties cannot be {@code Optional} or {@code null}.</em> It is expected that, in the event
> * that a group property can be disabled, this should be done using a {@code boolean} property (preferably the unnamed
> * property or a {@code enabled} property).

michael-simons commented on integration-tests/neo4j/README.md:
```diff
@@ -0,0 +1,21 @@
+
+Alternatively you can connect to your own Neo4j instance or cluster.
+Reconfigure the connection URL with `-Dneo4j.uri=bolt+routing://yourcluster:7687`;
+you'll probably want to change the authentication password too: `-Dneo4j.password=NotS0Secret`.
```
I don't understand? The `application.resources` in the integration test looks like this

```
quarkus.neo4j.uri = ${neo4j.uri}
quarkus.neo4j.authentication.username = ${neo4j.username}
quarkus.neo4j.authentication.password = ${neo4j.password}
```

and thus - if I understand your config mechanism correct - takes care of that for me.

But I'll add this as a note to the readme.

gsmet replied:
Ah sorry, missed that you used the Maven properties in the config file.

michael-simons commented:
All the suggestions applied, squashed all the polishing into one commit and rebased the whole thing on master again.

It's now based on the the latest Neo4j driver that doesn't expose the faulty packaging / shading of reactor anymore.

I'm looking forward to your opinion. I have some conversations above left open, in case you have any more questions.

gsmet commented on integration-tests/neo4j/src/test/java/io/quarkus/it/neo4j/Neo4jFunctionalityInGraalITCase.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Test various Neo4j operations running in SubstrateVM.
+ *
+ * @author Michael J. Simons
```
Remaining `@author` tag.

gsmet reviewed (changes requested):
I think we're nearly there.

I added some final comments and questions. I'm not sure about how the Netty event loops are consumed.

I just don't understand how the driver would consume the ones created by Quarkus?

michael-simons commented:
2nd round of polishing force pushed. Sorry for the author-tags, that's been an oversight.
Boss groups gone and so are all the maven opts. In fact, we don't need any of them.

gsmet commented on integration-tests/neo4j/pom.xml:
```diff
@@ -168,11 +169,7 @@
                                     <cleanupServer>true</cleanupServer>
-                                    <enableHttpUrlHandler>true</enableHttpUrlHandler>
-                                    <graalvmHome>${graalvmHome}</graalvmHome>
-                                    <enableJni>true</enableJni>
```
Hmmm, could you keep `<graalvmHome>${graalvmHome}</graalvmHome>`?

michael-simons commented:
In `quarkus/integration-tests/neo4j/src/main/resources/application.properties` I forgot to remove `quarkus.ssl.native=false`. 

So: Is this necessary for your CI? I'd love to remove it, because it will now disable encryption between server and client.

gsmet reviewed (changes requested):
Added one more comment. Other than that, it looks good.

Could you take care of it and squash everything?

Then let's wait for CI and get this merged!

Thanks for you work and your reactivity.

gsmet commented:
> So: Is this necessary for your CI? I'd love to remove it, because it will now disable encryption between server and client.

Yes, unfortunately.

michael-simons commented:
> Yes, unfortunately.

It would be super cool when CI would be able to pass this as environment variable. When configured like this in the application properties, that people might take as example, it promotes bad config.

gsmet commented:
@michael-simons so I guess you're volunteering to write a guide and a quickstart so that people don't take the IT as an example? O:)

https://quarkus.io/guides/
https://github.com/quarkusio/quarkus-quickstarts
:)

michael-simons commented:
Sure, will do. Not a thing (honestly).

But beforehand, I'll understand… why is it set to true here `/Users/msimons/Projects/misc/quarkus/integration-tests/camel-aws-sns/src/main/resources/application.properties`?

gsmet commented:
@michael-simons AFAIR, the Camel tests are not run on CI in native mode (all the Camel extensions will be moved soon outside of the tree).

michael-simons commented:
Ok, thanks. So I used `SslContextConfiguration.isSslNativeEnabled()` to detect whether native ssl is enabled or not.

Is there a way during runtimes to figure out if runtimes is actually native?

gsmet commented:
Yes, see `org.graalvm.nativeimage.ImageInfo`.

michael-simons commented:
That would be the thing to do during runtime, right?

"When native image and ssl native is not enabled…"


```
 if (ImageInfo.inImageRuntimeCode() && !SslContextConfiguration.isSslNativeEnabled()) {
            log.warn("Native SSL is disabled, communication between this client and the Neo4j server won't be encrypted.");
            configBuilder = configBuilder.withoutEncryption();
        }
```

michael-simons commented:
I'm quite sure we're ready now. The pom looks clean with the added GraalVM home and I'm happy with the warning logging regarding SSL. Thanks again for your input!

michael-simons commented:
FYI rebased onto master and squashed.

gsmet reviewed (approved):
Nice work, thanks!

I'm waiting for another PR to go in. Then I'm gonna merge this one.

BTW, could you write a short text presenting Neo4j, what use cases it is for and what this extension is doing? I will include it in the release blog post tomorrow.

gsmet commented:
And merged!

michael-simons commented:
Awesome!

I've already started work on the guide
See: https://github.com/michael-simons/quarkus/blob/feature/neo4j-guide/docs/src/main/asciidoc/neo4j-guide.adoc

Feel free to take the intro from there.

I'm looping in our chief devrel-caretaker @jexp for the intro.
